### PR TITLE
Mobile 2748

### PR DESCRIPTION
**Pull request by crazyserver**

### CODE REVIEW

File: src/core/features/course/classes/module-base-handler.ts [hunk 1]
```diff
@@ -34,24 +34,25 @@ export class CoreModuleHandlerBase implements Partial<CoreCourseModuleHandler> {
...
+    ): Promise<CoreCourseModuleHandlerData> {
         return {
-            icon: CoreCourse.getModuleIconSrc(module.modname, 'modicon' in module ? module.modicon : undefined),
+            icon: await CoreCourse.getModuleIconSrc(module.modname, 'modicon' in module ? module.modicon : undefined),
             title: module.name,
             class: 'addon-mod_' + module.modname + '-handler',
             showDownloadButton: true,
-            action: (event: Event, module: CoreCourseModule, courseId: number, options?: CoreNavigationOptions): void => {
+            mainPageName: this.pageName,
+            action(event: Event, module: CoreCourseModule, courseId: number, options?: CoreNavigationOptions): void {
```
Comment (dpalou) [inline src/core/features/course/classes/module-base-handler.ts:49]: Why did you change this back? When I reviewed the issue about the super class we agreed it was better to use arrow function in "action", that way you can use this.pageName in it and mainPageName isn't needed.

File: src/addons/mod/resource/services/handlers/module.ts [hunk 2]
```diff
@@ -164,28 +161,22 @@ export class AddonModResourceModuleHandlerService extends CoreModuleHandlerBase
...
             const mimetype = module.contentsinfo.mimetypes[0];
             if (mimetype) {
-                resourceData.icon = CoreMimetypeUtils.getMimetypeIcon(mimetype);
+                mimetypeIcon = CoreMimetypeUtils.getMimetypeIcon(mimetype);
             }
-            resourceData.extra = CoreTextUtils.cleanTags(module.afterlink);
+            extra.push(CoreTextUtils.cleanTags(module.afterlink));
 
-        } else if (files && files.length) {
+        } else if (module.contents[0]) {
```
Comment (dpalou) [inline src/addons/mod/resource/services/handlers/module.ts:175]: I think it's better to make sure that module.contents exists, like it was done before, since it can be undefined. They should have been loaded in hideOpenButton, but that could change in the future  so IMO it's better to do: module.contents && module.contents[0]

File: src/addons/mod/url/services/handlers/module.ts [hunk 3]
```diff
@@ -118,10 +118,11 @@ export class AddonModUrlModuleHandlerService extends CoreModuleHandlerBase imple
 
             handlerData.buttons[0].hidden = hideButton;
 
-            if (module.contents && module.contents[0]) {
-                // Calculate the icon to use.
-                handlerData.icon = await CoreCourse.getModuleIconSrc(module.modname, module.modicon) ||
-                    AddonModUrl.guessIcon(module.contents[0].fileurl);
+            if (module.contents[0]) {
```
Comment (dpalou) [inline src/addons/mod/url/services/handlers/module.ts:121]: Like above, I think it's better to keep the previous check to prevent future errors.

File: src/core/features/course/services/course.ts [hunk 4]
```diff
@@ -1484,12 +1489,12 @@ export type CoreCourseWSModule = {
     noviewlink?: boolean; // Whether the module has no view page.
     completion?: number; // Type of completion tracking: 0 means none, 1 manual, 2 automatic.
     completiondata?: CoreCourseModuleWSCompletionData; // Module completion data.
-    contents?: CoreCourseModuleContentFile[];
+    contents: CoreCourseModuleContentFile[];
```
Comment (dpalou) [inline src/core/features/course/services/course.ts:1492]: This change is wrong. contents can be undefined, e.g. if the module has an access restriction.

File: src/core/features/course/services/course.ts [hunk 5]
```diff
@@ -402,29 +402,30 @@ export class CoreCourseProvider {
         ): Promise<CoreCourseWSSection[]> => {
             const params: CoreCourseGetContentsParams = {
                 courseid: courseId!,
-                options: [],
             };
+            params.options = [];
```
Comment (dpalou) [inline src/core/features/course/services/course.ts:406]: Is this change needed? Is TS unable to detect that options exist if they're in the initial object?

File: src/addons/mod/resource/services/handlers/module.ts [hunk 6]
```diff
@@ -94,7 +93,7 @@ export class AddonModResourceModuleHandlerService extends CoreModuleHandlerBase
         }];
 
         this.getResourceData(module, courseId, handlerData).then((data) => {
-            handlerData.icon = data.icon;
+            handlerData.icon = data.icon; // @TODO
```
Comment (dpalou) [inline src/addons/mod/resource/services/handlers/module.ts:96]: What is this TODO about?

Also, I think the function getResourceData doesn't return an icon right now, and I don't see it changed in your patch.

File: src/addons/mod/url/services/handlers/module.ts [hunk 7]
```diff
@@ -111,17 +111,18 @@ export class AddonModUrlModuleHandlerService extends CoreModuleHandlerBase imple
...
             handlerData.buttons[0].hidden = hideButton;
 
-            if (module.contents && module.contents[0]) {
-                // Calculate the icon to use.
-                handlerData.icon = AddonModUrl.guessIcon(module.contents[0].fileurl) ||
-                    CoreCourse.getModuleIconSrc(module.modname, 'modicon' in module ? module.modicon : undefined);
+            if (module.contents[0]) {
+                const icon = AddonModUrl.guessIcon(module.contents[0].fileurl);
+
+                // Calculate the icon to use. @TODO
```
Comment (dpalou) [inline src/addons/mod/url/services/handlers/module.ts:124]: Should this TODO be here? Aren't you calculating the icon already?

File: src/core/features/course/services/course.ts [hunk 8]
```diff
@@ -632,7 +631,11 @@ export class CoreCourseProvider {
      * @param modicon The mod icon string to use in case we are not using a core activity.
      * @return The IMG src.
      */
-    getModuleIconSrc(moduleName: string, modicon?: string): string {
+    async getModuleIconSrc(moduleName: string, modicon?: string, mimetypeIcon = ''): Promise<string> {
+        if (mimetypeIcon) {
+            return mimetypeIcon;
+        }
+
```
Comment (dpalou) [inline src/core/features/course/services/course.ts:638]: I see that the modicon is only returned if the module isn't supported in the app. Does this mean that for core icons we'll always use the local icons? I thought we wanted to use the remote icons.

**Review by dpalou — CHANGES_REQUESTED** [review]

**Comment by dpalou** [comment]
One thing I forgot to mention is that in some places you're passing componentId, so the icons will be linked to that instance. This means that removing the stored files for that instance will delete the icon. This is right for unique icons (e.g. LTI with a custom icon), but generic icons shared by all instances shouldn't be deleted if an instance data is deleted.

File: src/core/features/course/classes/module-base-handler.ts (hunk 1 above)
Comment (crazyserver) [reply src/core/features/course/classes/module-base-handler.ts:49]: I think it's a rebase error, reverted now.

File: src/addons/mod/resource/services/handlers/module.ts (hunk 2 above)
Comment (crazyserver) [reply src/addons/mod/resource/services/handlers/module.ts:175]: Checking CoreCourseWSModule contents is always set. That's why I changed it.

Comment (crazyserver) [reply src/addons/mod/resource/services/handlers/module.ts:175]: Ok, reverting it

File: src/core/features/course/services/course.ts (hunk 4 above)
Comment (crazyserver) [reply src/core/features/course/services/course.ts:1492]: Ah ok, I see the problem now.

File: src/core/features/course/services/course.ts (hunk 5 above)
Comment (crazyserver) [reply src/core/features/course/services/course.ts:406]: It says Object possibly undefined.

File: src/addons/mod/url/services/handlers/module.ts (hunk 7 above)
Comment (crazyserver) [reply src/addons/mod/url/services/handlers/module.ts:124]: no, deleted.

File: src/addons/mod/resource/services/handlers/module.ts (hunk 6 above)
Comment (crazyserver) [reply src/addons/mod/resource/services/handlers/module.ts:96]: TODO deleted, getResourceData is returning icon, probably you misread some code.

**Comment by crazyserver** [comment]
Thanks Dani! Ready to review again!

**Review by dpalou — APPROVED** [review]
